[CI] [GHA] Introduce macOS ARM64 as a matrix parameter in the macOS pipeline

akashchi opened the pull request:
### Tickets:
 - *122475*

akashchi commented on .github/workflows/mac.yml:
```diff
@@ -98,11 +109,12 @@ jobs:
           # Should save cache only if run in the master branch of the base repo
           # github.ref_name is 'ref/PR_#' in case of the PR, and 'branch_name' when executed on push
-          save: ${{ github.ref_name == 'master' && 'true' || 'false'  }}
+          # TODO: Uncomment
+          # save: ${{ github.ref_name == 'master' && 'true' || 'false'  }}
```
Self-reminder

ilya-lavrenov commented on .github/workflows/mac.yml:
```diff
@@ -34,11 +34,22 @@ jobs:
       CMAKE_BUILD_TYPE: 'Release'
       CMAKE_GENERATOR: 'Ninja Multi-Config'
-      MACOSX_DEPLOYMENT_TARGET: '10.12'
+      MACOSX_DEPLOYMENT_TARGET: ${{ matrix.arhitecture == 'x86_64' && '10.12' || '11.0' }}
```
may be we can move it to matrix to see all the difference between pipelines...

ilya-lavrenov commented on tools/openvino_dev/setup.py:
```diff
@@ -48,7 +48,7 @@
         'black_list': [],
         'prefix': 'omz_tools',
-        'extract_requirements': True,
+        'extract_requirements': False,
```
```suggestion
        'extract_requirements': True,
```

ilya-lavrenov commented on .github/workflows/mac.yml:
```diff
@@ -482,17 +546,19 @@ jobs:
           # Find and install OV dev wheel
           pushd ${{ env.INSTALL_DIR }}/tools
             ov_dev_wheel_name=$(find . -name 'openvino_dev*.whl')
-            python3 -m pip install $ov_dev_wheel_name[mxnet,caffe,kaldi,onnx,tensorflow2]
```
note, that mxnet is available only for macos x86_64
explicitly removed this extras in https://github.com/openvinotoolkit/openvino/pull/20403

So, we need conditional install depending on architecture (and similar for tests below)

akashchi commented:
I am in the process of skipping the C++ and Python tests, the skips should be ready soon.

ilya-lavrenov commented on tests/layer_tests/mo_python_api_tests/test_mo_convert_pytorch.py:
```diff
@@ -1236,6 +1237,7 @@ class TestPrecisionSensitive():
     @pytest.mark.parametrize("create_model", test_data)
     @pytest.mark.nightly
     @pytest.mark.precommit
+    @pytest.mark.skipif(platform.system() == 'Darwin' and platform.machine() == 'arm64', reason='Ticket - 122714')
```
I suppose `xfail` is better, because it will fail if test is fixed

akashchi commented:
I guess this one would need a touch of refactoring after #20388 is merged.

akashchi commented:
Seems like there are a couple more failings after merging master, I will get to skipping them shortly.

ilya-lavrenov commented on .github/workflows/mac.yml:
```diff
@@ -1,14 +1,14 @@
-name: macOS (macOS 12, Python 3.11)
+name: macOS (macOS 13, Python 3.11, x86_64 and ARM64)
```
```suggestion
name: macOS (Python 3.11)
```
because x86_64 or arm64 will be added by matrix

ilya-lavrenov commented on .github/workflows/mac.yml:
```diff
@@ -34,11 +34,24 @@ jobs:
+      fail-fast: false
+      matrix:
+        include:
+          - name: 'macOS x86_64'
```
```suggestion
```
because `x86_64` will be added by `arhitecture`

ilya-lavrenov commented on src/frontends/onnx/tests/tests_python/test_backend.py:
```diff
@@ -683,6 +685,22 @@ def expect_fail(test_case_path, xfail):  # type: (str) -> None
+if platform.system() == 'Darwin':
+    tests_expected_to_fail.extend([
+        (
+            skip_issue_58676,
```
why is it skip_issue_58676 and not xfail_issue_58676 ?

akashchi replied:
Seems like it is inconsistent on macOS as we have seen on master with #20367 and #20402: one pipeline was okay, but the next was failing in the post-commit because of the test accidentally passing and failing the pipeline with `xpass`.

akashchi commented:
The pipelines are passing: https://github.com/openvinotoolkit/openvino/actions/runs/6610350804?pr=20363. I've removed the PR and post-commit triggers and added a nightly trigger as it is in the Windows pipeline as ARM Macs are even pricier than x86_64 Macs.
